XDMA documentation (#206)
KULeuven-MICAS/snax_cluster: hw/chisel/doc/xdma.md

  - [Source Code Hierarchy](#source-code-hierarchy)
  - [Features](#features)
    - [New Data-fetching Unit (Streamer) to Support Powerful Direct Memory Access](#new-data-fetching-unit-streamer-to-support-powerful-direct-memory-access)
    - [Extendable Datapath Extension to Enable *Computation on Copy* Capability](#extendable-datapath-extension-to-enable-computation-on-copy-capability)
    - [Distributed Datapath for Cluster-to-Cluster or Host-to-Cluster Memory Access](#distributed-datapath-for-cluster-to-cluster-or-host-to-cluster-memory-access)
  - [CSR Address Definition](#csr-address-definition)
  - [Design of Custom Datapath Extension](#design-of-custom-datapath-extension)
    - [Native Datapath Extension](#native-datapath-extension)
    - [SystemVerilog Datapath Extension](#systemverilog-datapath-extension)
  - [Integration with SNAX Platform](#integration-with-snax-platform)
    - [Hardware Interface](#hardware-interface)
    - [Software Interface](#software-interface)

## Introduction and XDMA hardware block diagram

XDMA is implemented in every cluster with the following architecture:

> ![XDMA Hardware Diagram](xdma_figures/xdma_datapath.png)

Reviewer: So that you don't need to save pictures, you can copy-paste and paste them directly to the github page! So it goes like:

1. Take a screen shot of your image.
2. Paste it here directly on the .md in github.
3. Github creates its own link to the image.

Just a suggestion. I was told to do that better than to save pictures here all the time.

Reviewer 2: yes this is better

Author: Haha okay! I wrote the markdown file offline (the preview renderring happens on my computer locally), so I don't know github natively supports it...

> At the top of the diagram is the interface to the TCDM. IO operations towards memory occur here. The Address Generation Unit generates addresses consumed by each channel. The accessed data is forwarded to XDMA datapath extension chains, undergoing several programmable processes. The data is either looped back to the writer side of the same XDMA if the destination is in the same region as the source, or sent to the writer side of a remote XDMA if in a different region. The writer's side mirrors the reader's side but inverts the data flow direction.

Reviewer 2: ```suggestion
At the top of the diagram is the interface to the TCDM. IO operations towards local memory occur here. The interface consists of multiple channels accessing data in parallel. The Address Generation Unit generates addresses consumed by each channel. The accessed data is forwarded to XDMA datapath extension chains, undergoing several programmable processes. The data is either looped back to the writer side of the same XDMA if the destination is in the same region as the source, or sent to the writer side of a remote XDMA if in a different region. The writer's side mirrors the reader's side but inverts the data flow direction.
```

>   - **DataResponser.scala**: Module receiving responses from the TCDM interface.
>   - **Reader.scala**: Reads data from the TCDM, incorporating both a **DataRequestor** and a **DataResponser**.
>   - **Writer.scala**: Writes data to the TCDM, incorporating a **DataRequestor**.
> - **xdmaTop**
>   - **xdmaTop.scala** The top module and entry point of the entire Scala (Chisel) program, acting as a SystemVerilog Generator.

## Features

This section introduces three key features of XDMA:

### New Data-fetching Unit (Streamer) to Support Powerful Direct Memory Access

The TCDM interconnect at the top of the figure provides full-region data access with a one-cycle delay if there is no contention. XDMA's configuration includes variable bit widths for both memory and AXI channels. The number of data streamers on both reader and writer sides is automatically determined by these configurations, e.g.,:

- **TCDMDataWidth=64** and **AXIDataWidth=512**: Typical for Snax clusters. XDMA will instantiate eight channels for both reading and writing at the TCDM side, concatenating them into a single 512-bit channel at the AXI side.

> - **TCDMDataWidth=64** and **AXIDataWidth=64**: Typical for global scratchpad memory/SRAM. XDMA will instantiate only one channel on each side, adopting a simpler architecture used by most other DMAs.

Reviewer 2: How does this look for L3 memory in the snax cluster? I thought this would be `**TCDMDataWidth=512** and **AXIDataWidth=512**`

Author: Yes. Another way is to use original idma as the "remote" xdma, which has strong affinity to L3 memory.

> Each TCDM channel has its own address and data caches, enabling independent data fetching with minimal interference from other channels. Channels can be selectively disabled to save energy and simplify memory initialization (e.g., turning off all reader channels and enabling writer channels to write zeros to memory). The byte mask is also implemented for byte-level granularity in writing.

Reviewer: `data caches` it has?

Author: Precisely it is a FIFO. But acting as the buffer.

> Each TCDM channel has its own address and data caches, enabling independent data fetching with minimal interference from other channels. Channels can be selectively disabled to save energy and simplify memory initialization (e.g., turning off all reader channels and enabling writer channels to write zeros to memory). The byte mask is also implemented for byte-level granularity in writing.
> 

Reviewer 2: ```suggestion
Each TCDM channel has its own address and data caches, enabling independent data fetching with minimal interference from other channels. Channels can be selectively disabled if the required bandwidth is less than the bandwidth of all channels combined (e.g., turning off all reader channels and enabling writer channels to write zeros to memory). Additionally, a byte mask is also implemented for byte-level granularity in writing the data back to memory.

```
The goal of disabling channels is not to save energy I think?

Author: You are right indeed. What I want to say here is save more energy compared with masking😂😂

The Datapath Extension is a critical component allowing on-the-fly data processing such as matrix transposing, max-pooling, average-pooling, and data stream compression/decompression. By supplying the necessary parameters to the XDMA generator, extension modules are instantiated sequentially, with each module's output feeding into the next module's input.

Designing a custom datapath extension is straightforward thanks to Chisel's flexibility. The pipeline is automatically inserted, CSR configs are automatically wired, infrastructure for enabling/disabling extensions is added, and the corresponding C header for runtime is automatically generated. Integrating native extensions written in Chisel or extensions written in SystemVerilog is remarkably simple.

### Distributed Datapath for Cluster-to-Cluster or Host-to-Cluster Memory Access

XDMA can perform data copying within the same memory region or across different regions. If the source and destination addresses are from the same region, data is looped back directly without utilizing the AXI, conserving energy. If addresses span different regions, the data is handled in three steps:

> 1. The controller recognizes that the source data is from a non-local region and flattens the configuration into a one-dimensional frame.

Reviewer 2: ```suggestion
1. The controller recognizes that the source data is from a non-local region and flattens the configuration into a one-dimensional frame with a width equal to the bitwidth of the AXI bus.
```

2. The configuration frame is sent to the remote XDMA to request the data.

> 3. The data is relayed back from the remote to the local site.

Reviewer 2: Can you request data from a remote XDMA or only send data to a remote XDMA or both?

Author: We can only request the data from the remote. Write to TCDM only happens at the local. Choose not to support both condition is because the synchronization can be a problem, and it potentially become the cause of deadlock.

An illustration of the XDMA call chain is shown below:

> ![XDMA Call Chain](xdma_figures/xdma_callchain.png)

Reviewer 2: what is B0 / B1?

Author: The figure is updated

Note: The AXI protocol adaptor is still under development and the architecture may change.

## CSR Address Definition

Consider the following configuration:

```hjson
// Key configuration items only:

  // CFG items inside cluster
  cluster: {
    addr_width: 48,
    data_width: 64,
    dma_data_width: 512
  },
  // CFG items inside XDMA core template
  xdma_core_template: {
     snax_xdma_cfg: {
         reader_buffer: 8, 
         writer_buffer: 8, 
         reader_agu_dimension: 3, 
         writer_agu_dimension: 3, 
         HasTransposer: 3, 
         HasMemset: 1, 
         HasMaxPool: 2
     }
  }
```

This configuration specifies eight channels for both reader and writer TCDM ports, employing three-dimensional address generation. The innermost dimension is used explicitly by the eight parallel channels, and its value can be reduced to disable some channels. The other two dimensions are managed by the Address Generation Unit (AGU).

> The enabled extensions — **HasTransposer**, **HasMemset**, and **HasMaxPool** — add bypass signals and custom CSRs to the configuration list. The positive number at the back means it will be enabled, and the size of the number indicate its sequence - the one with smaller value will in the front of one with larger value. 

Reviewer 2: this is cool!

Author: Haha the effort to make the "config file" the only source of truth is tremendous. Imagine the case that the users add the new extension. If after writing the actual implementation, the users still need to modify generator to do the correct generation of the new extension. It is not elegant at all, and this asks the users to understand scala before they can do integration. To address this, I use meta programming provided by Scala (and Java), which means while Chisel generates SystemVerilog, Chisel also generates Chisel itself. See #207 for more information.

The following table outlines the CSR addresses and their settings:

> | CSR Address | Privilege |                Description               |      Effective Bits / Value      |

Reviewer: Hmm will the XDMAs be fixed to have the CSR addressing to start at 960? Or should it be relative addressing based on where we want it to start?

Author: XDMA reuses CSR manager from SNAX, so the implemenation of CSR manager in SNAX decides that I cannot change the CSR start address... 

Is it possible to add a parameter in CSR manager to make it modifiable?

|:-----------:|:---------:|:----------------------------------------:|:--------------------------------:|
|     960     |     RW    |      The LSB of Source Base Address      |              [31:0]              |

> |     961     |     RW    |      The MSB of Source Base Address      |              [15:0]              |

Reviewer 2: oh no do we have to use 48 bit addresses in occamy?

Author: Yes... Occamy uses 48b address. I think it is good, as there is more address spaces for the future chiplet version

|     977     |     RW    |    Bypass Enabler of  Memset Extension   |               0, 1               |
|     978     |     RW    |        Memset Extension Custom CSR       |               [7:0]              |
|     979     |     RW    |    Bypass Enabler of MaxPool Extension   |               0, 1               |
|     980     |     RW    |       MaxPool Extension Custom CSR       |               [7:0]              |
|     981     |     RW    |   Bypass Enabler of Transpose Extension  |               0, 1               |
|     982     |     RW    |       Commit Current Configuration       |                 1                |
|     983     |     R     |      Committed Configuration Counter     |              [31:0]              |
|     984     |     R     |      Finished Configuration Counter      |              [31:0]              |

Offsets may vary based on the configuration, but the Chisel code will automatically generate the necessary C headers, ensuring ease of use for developers.

## Design of Custom Datapath Extension

### Native Datapath Extension

Designing a Datapath Extension involves four steps:

> 

Reviewer 2: Do the extensions have a handshake with the input data or should it be a 0-cycle delay datapath?

Author: There is handshaking. You can design it in full-combinational style, but you can also add arbitrary latency to it, as there is handshake signal. 

MaxPool extension is with the latency. Furthermore, the input length and output length are not equal. XDMA also supports it. 

Transposer extension is with no latency. However, XDMA will still add one delay implicitly, to prevent critical path being too long after the concatenation of many combinational extension.
